pkg/envoy/accesslog/request_header_operator.go

```diff
@@ -0,0 +1,84 @@
+package accesslog
+
+import (
+	"fmt"
+
+	accesslog_config "github.com/envoyproxy/go-control-plane/envoy/config/accesslog/v2"
+	accesslog_data "github.com/envoyproxy/go-control-plane/envoy/data/accesslog/v2"
+
+	envoy_core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+)
+
+// RequestHeaderOperator represents a `%REQ(X?Y):Z%` command operator.
+type RequestHeaderOperator struct {
+	HeaderFormatter
+}
+
+func (f *RequestHeaderOperator) FormatHttpLogEntry(entry *accesslog_data.HTTPAccessLogEntry) (string, error) {
+	return f.Format(&RequestHeaders{entry.GetRequest()})
+}
+
+func (f *RequestHeaderOperator) FormatTcpLogEntry(entry *accesslog_data.TCPAccessLogEntry) (string, error) {
+	return "", nil
+}
+
+func (f *RequestHeaderOperator) ConfigureHttpLog(config *accesslog_config.HttpGrpcAccessLogConfig) error {
+	config.AdditionalRequestHeadersToLog = f.AppendTo(config.AdditionalRequestHeadersToLog)
+	return nil
+}
+
+func (f *RequestHeaderOperator) ConfigureTcpLog(config *accesslog_config.TcpGrpcAccessLogConfig) error {
+	// has no effect on TcpGrpcAccessLogConfig
+	return nil
+}
+
+// String returns the canonical representation of this access log fragment.
+func (f *RequestHeaderOperator) String() string {
+	return fmt.Sprintf("%%REQ%s%%", f.HeaderFormatter.String())
+}
+
+// RequestHeaders represents a set of HTTP request headers
+// that includes both regular headers, such as `referer` and `user-agent`,
+// and pseudo headers, such as `:method`, `:authority` and `:path`.
+type RequestHeaders struct {
+	*accesslog_data.HTTPRequestProperties
+}
+
+func (h *RequestHeaders) Get(name string) (string, bool) {
+	switch name {
```

Comment:
No, `HTTPAccessLogEntry.Request.RequestHeaders` only includes those headers that are configured explicitly in the config of `envoy.http_grpc_access_log`.

See [HttpGrpcAccessLogConfig. additional_request_headers_to_log](https://github.com/envoyproxy/envoy/blob/v1.13.0/api/envoy/config/accesslog/v2/als.proto#L28-L30) and [Envoy sources](https://github.com/envoyproxy/envoy/blob/v1.13.0/source/extensions/access_loggers/grpc/http_grpc_access_log_impl.cc#L113-L122).

Please find below a dump of access log message sent out of `Envoy`.

Notice that values such as `authority`, `path`, etc are sent out as individual Protobuf fields rather than a map.

```yaml
httpLogs:
  logEntry:
  - commonProperties:
      downstreamDirectRemoteAddress:
        socketAddress:
          address: 127.0.0.1
          portValue: 59145
      downstreamLocalAddress:
        socketAddress:
          address: 127.0.0.1
          portValue: 3003
      downstreamRemoteAddress:
        socketAddress:
          address: 127.0.0.1
          portValue: 59145
      startTime: "2020-02-24T11:09:51.654444Z"
      timeToFirstDownstreamTxByte: 0.001248848s
      timeToFirstUpstreamRxByte: 0.001220639s
      timeToFirstUpstreamTxByte: 0.000045517s
      timeToLastDownstreamTxByte: 0.001251200s
      timeToLastRxByte: 0.000012779s
      timeToLastUpstreamRxByte: 0.001220788s
      timeToLastUpstreamTxByte: 0.000049784s
      upstreamCluster: backend
      upstreamRemoteAddress:
        socketAddress:
          address: 127.0.0.1
          portValue: 18181
    protocolVersion: HTTP11
    request: # notice that the following values are sent out of Envoy as individual Protobuf fields rather than a map 
      authority: localhost:3003
      path: /145
      requestHeadersBytes: "188"
      requestId: de19e8ff-dbfa-414c-ba2e-dd0ca88f3278
      requestMethod: GET
      scheme: http
      userAgent: curl/7.64.1
    response:
      responseCode: 404
      responseCodeDetails: via_upstream
      responseHeadersBytes: "99"
```